### DISCUSSION TITLE
Apply highlight on the best match of the field - disable highlight on stop-words

### DISCUSSION THREAD

**Discussion by curquiza** [meilisearch/product#701] [Feedback & Feature Proposal]
Following this issue: https://github.com/meilisearch/meilisearch/issues/4105

A solution could be to provide a setting not to highlight the stop words

**Comment by curquiza** [comment]
Pining @ManyTheFish to let you know

Is disabling highlight on stop-words something possible to do?

**Reply by ManyTheFish** [reply]
We could, but it would lead to some unexpected outcomes as I said in this issue:
https://github.com/meilisearch/meilisearch/issues/4105#issuecomment-1764819747

> currently, Meilisearch highlights the stop words because we don't want the end-user to have a partial highlighting about this, for example, if you search for "soup of the day" and `of` and `the` are considered stop words, you don't want only to highlight `soup` and `day` like  "***soup*** of the ***day***".
However, the highlight flooding you are experiencing is indeed more disturbing than helpful, this could be enhanced by only highlighting the best matches per field, this would remove all this noise without having to ignore stopwords but it will probably miss some relevant matches in the process.

**Reply by curquiza** [reply]
I read your comment but I'm not sure to understand what are the unexpected outcome you mention.

> if you search for "soup of the day" and of and the are considered stop words, you don't want only to highlight soup and day like "soup of the day".

If users disable stop words to be highlighted, they will accept this kind of behavior IMO. How do you know the users "don't want" to only highlight soup and day? Real question here, because I guess I missed something

**Reply by ManyTheFish** [reply]
The end-user and the user choosing the settings (the administrator) are not the same, which means that if an Administrator chooses to put `the` and `of` as stop words, and, an end-user searches for `soup of the day`, the retrieved document will be highlighted as:
```
For starters, you can choose between a smoothie made from Lanaudière berries or the <em>soup</em> of the <em>day</em>.
```

which could seem weird from the end-user POV who doesn't know the stop words and their behavior. 🤔

**Reply by curquiza** [reply]
It's the same for all the settings: the administrator can impact the relevancy or anything via settings and the end user could expect something different. It's up to the developers to define the settings they need because they better know their own final users.

> which could seem weird from the end-user POV who doesn't know the stop words and their behavior. 🤔

It could seem weird for some usecases. But it can be something they want for other usecases. The developers know better their final users than we do. For example, even if it's a personal feeling, it does not seem weird to me to only highlight `soup` and `day` if you type `soup of the day`. That's why I did not really understand your answer first. But again a really personal opinion.

That's why I suggested first this option: by default we highlight stop words, and we could let the users activate a setting to not highlight them if they want. Not sure this is the best solution, but a suggestion.

Since it's mostly about opinion here, about what is weird or not, we need the point of view of users before making any decision. Maybe some of them don't want to highlight the stop words, as the users described it in the issue we linked.

**Reply by abdullah-alnahas** [reply]
I think a good way to implement this is by introducing an additional parameter to specify whether or not to highlight stop words. This offers more flexibility than placing it in the settings. What do you think?

**Reply by curquiza** [reply]
@ManyTheFish I think I definitely do not understand something here. Not highlighting stop-words is already the by default Meilisarch behavior:

with stop words `of` and `the`

<img width="1292" alt="Capture d’écran 2023-10-18 à 15 21 22" src="https://github.com/meilisearch/product/assets/20380692/25ab015d-6fd9-41b3-a4f1-de34a0867392">


without any stop words

<img width="1395" alt="Capture d’écran 2023-10-18 à 15 24 35" src="https://github.com/meilisearch/product/assets/20380692/44457ec3-aa74-490f-a4ba-2171a7271ee8">

**Reply by abdullah-alnahas** [reply]
> Not highlighting stop-words is the default behavior in Meilisearch.

@curquiza, do you need to update the stop words set (and add words like 'the') when you create a new index? Or is it populated with a default set for each language?

**Reply by curquiza** [reply]
You have to update your stop words yourself @abdullah-alnahas, Meilisearch will not add default stop-words depending on your dataset language

**Reply by ManyTheFish** [reply]
@curquiza, my bad, indeed the stop word are not highlighted, the highlighter doesn't get the stop words from the search query, they have been filtered before. This means that the stopword shouldn't be highlighted. It's seems to be a bug, sorry again it's my bad.
More precisely about the source of this behavior, I don't know why the `"ال"` is highlighted.

**Comment by curquiza** [comment]
@abdullah-alnahas can you described a little bit more your usage. What is problematic for you currently?
Is it the same issue than https://github.com/meilisearch/meilisearch/issues/4105 or different?

We need to be sure adding the setting is the right solution to fix the issue

**Reply by abdullah-alnahas** [reply]
Yes, it is the same issue. As I replied [above](https://github.com/meilisearch/product/discussions/701#discussioncomment-7314026), perhaps introducing a parameter instead of a setting would offer more flexibility. What are your thoughts on this?

**Comment by curquiza** [comment]
Closing the discussion since it's not a feature request, but [probably a bug](https://github.com/orgs/meilisearch/discussions/701#discussioncomment-7421540). We will re-open a discussion later if needed, but the base of the discussion here are not good
